File: 139.WordBreak.cpp
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        unordered_set<string> set(wordDict.begin(), wordDict.end());
        vector<bool> dp(s.size()+1);
        dp[0] = true;
        for (int i=0; i<=s.size(); ++i) {
            for (int j=0; j<i; ++j) {
                if (dp[j] && set.count(s.substr(j, i-j))) {
                    dp[i] = true; break;
                }
            }
        }
        return dp.back();
    }
}

File: 139.WordBreak.java
class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        Set<String> dict = new HashSet<String>(wordDict);
        boolean[] dp = new boolean[s.length()+1];
        dp[0] = true;
        for (int i=1; i<=s.length(); ++i) {
            for (int j=0; j<i; ++j) {
                if (dp[j] && dict.contains(s.substring(j, i))) {
                    dp[i] = true; break;
                }
            }
        }
        return dp[s.length()];
    }
}
